unified-runtime/source/adapters/offload/CMakeLists.txt

```diff
@@ -5,14 +5,84 @@
 
 set(TARGET_NAME ur_adapter_offload)
 
-set(UR_OFFLOAD_INSTALL_DIR "" CACHE PATH "Path to the directory containing libomptarget.so etc")
-if (UR_OFFLOAD_INSTALL_DIR STREQUAL "")
-    message(FATAL_ERROR "UR_OFFLOAD_INSTALL_DIR must be defined for the Offload adapter")
-endif()
+add_ur_adapter(${TARGET_NAME}
+    SHARED
+    ${CMAKE_CURRENT_SOURCE_DIR}/adapter.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/context.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/device.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/enqueue.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/event.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/kernel.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/memory.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/platform.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/program.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/queue.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/ur2offload.hpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/ur_interface_loader.cpp
+    ${CMAKE_CURRENT_SOURCE_DIR}/usm.cpp
+)
 
+set(UR_OFFLOAD_INSTALL_DIR "" CACHE PATH "Path to the directory containing libomptarget.so etc")
 set(UR_OFFLOAD_INCLUDE_DIR "" CACHE PATH "Path to the directory containing LLVM headers")
-if (UR_OFFLOAD_INCLUDE_DIR STREQUAL "")
-    message(FATAL_ERROR "UR_OFFLOAD_INCLUDE_DIR must be defined for the Offload adapter")
+if (UR_OFFLOAD_INSTALL_DIR STREQUAL "" OR UR_OFFLOAD_INCLUDE_DIR STREQUAL "")
+    include(ExternalProject)
+
+    set(LLVM_PROJECT_GIT https://github.com/llvm/llvm-project.git)
+    set(LLVM_PROJECT_TAG ffddf33beb8097e760fafcbdd28d56c1d055e820)
+    message(STATUS "UR_OFFLOAD_INSTALL_DIR and/or UR_OFFLOAD_INCLUDE_DIR not set, building liboffload runtime from source using llvm-project tag: ${LLVM_PROJECT_TAG}")
+    set(LLVM_PROJECT_SOURCE_DIR ${CMAKE_BINARY_DIR}/llvm-project-src)
+    set(OPENMP_INSTALL_DIR ${CMAKE_BINARY_DIR}/openmp-install)
+    # Install to the same directory as main build
+    set(UR_OFFLOAD_INSTALL_DIR ${CMAKE_INSTALL_PREFIX})
+    set(UR_OFFLOAD_INCLUDE_DIR ${CMAKE_INSTALL_PREFIX}/include)
+    # Clone llvm-project only once
+    ExternalProject_Add(llvm_project_src
+        GIT_REPOSITORY ${LLVM_PROJECT_GIT}
+        GIT_TAG ${LLVM_PROJECT_TAG}
+        SOURCE_DIR ${LLVM_PROJECT_SOURCE_DIR}
+        CONFIGURE_COMMAND ""
+        BUILD_COMMAND ""
+        INSTALL_COMMAND ""
+        UPDATE_COMMAND ""
+    )
+    # Build OpenMP runtime (required dependency for offload's libomptarget) from the cloned source
+    ExternalProject_Add(openmp_ext
```

Comment:
Why are we not just using `LLVM_ENABLE_RUNTIMES="offload;openmp"` here?

Reply:
We’re not using `LLVM_ENABLE_RUNTIMES="offload;openmp"` here because that mode requires configuring from the top-level llvm directory. In that flow, clang/llvm would be built as part of the runtimes build, which means clang/llvm ends up being built twice.
Instead, I build OpenMP and offload in standalone mode, reusing the existing clang/llvm build from intel/llvm.

Also, in the latest commit I improved this further: rather than cloning llvm-project again, I now create a git worktree from the already cloned intel/llvm repository, pulling in only the directories needed for the runtime build.

Reply:
I looked into this a while ago and figured out how to build openmp without a dependency on clang. It required a few patches, which I should really figure out who to harass about.

> rather than cloning llvm-project again, I now create a git worktree from the already cloned intel/llvm repository

What if the runtimes are using a newer version of llvm-project that hasn't been pulled down into intel/llvm yet?